libr/core/dmh_glibc.inc.c

```diff
@@ -100,32 +100,231 @@ static GHT GH(get_main_arena_with_symbol)(RCore *core, RDebugMap *map) {
+			version = strtod (version_start, NULL);
+		}
+		r_list_free (matches);
+		r_regex_free (rx);
+	}
+
+	if (version != 0) {
+		R_LOG_INFO ("libc version %.2f identified from .rodata banner", version);
+		return version;
 	}
+
+	R_LOG_WARN ("get_glibc_version failed");
+	return version;
+}
+
+static bool GH(resolve_glibc_version)(RCore *core) {
+	r_return_val_if_fail (core && core->dbg && core->dbg->maps, false);
+
+	double version = 0;
 	RDebugMap *map;
 	RListIter *iter;
+	bool found_glibc_map = false;
+
+	if (core->dbg->glibc_version_resolved) {
+		return true;
+	}
+
+	const char *dbg_glibc_version = r_config_get (core->config, "dbg.glibc.version");
+	if (R_STR_ISEMPTY (dbg_glibc_version)) {
+		dbg_glibc_version = NULL;
+	}
+
+	if (dbg_glibc_version)	{
+		// TODO: use ^ and $ which appear to be broken
+		if (!r_regex_match ("\\d.\\d\\d", "e", dbg_glibc_version)) {
+			R_LOG_WARN ("resolve_glibc_version: Unexpected version format in dbg.glibc.version: %s"
+						" (expected format \"\\d.\\d\\d\")", dbg_glibc_version);
+		} else {
+			version = strtod(dbg_glibc_version, NULL);
+			core->dbg->glibc_version = (int) round((version * 100));
+			core->dbg->glibc_version_d = version;
+			core->dbg->glibc_version_resolved = true;
+			R_LOG_INFO ("libc version %.2f set from dbg.glibc.version", core->dbg->glibc_version_d);
+			return true;
+		}
+	}
+
+	r_return_val_if_fail (core && core->dbg && core->dbg->maps, false);
 	r_debug_map_sync (core->dbg);
+
+	// Search for binary in memory maps named *libc-* or *libc.*
+	// TODO: This is very brittle, other bin names or LD_PRELOAD could be a problem
 	r_list_foreach (core->dbg->maps, iter, map) {
-		// In case the binary is named *libc-* or *libc.*
-		if (strncmp (map->name, core->bin->file, strlen (map->name)) != 0) {
-			char *fp = strstr (map->name, "libc-");
-			if (fp) {
-				break;
-			}
-			fp = strstr (map->name, "libc.");
-			if (fp) {
-				v = r_num_get_double (core->num, fp + 5);
-				core->dbg->glibc_version = (int) round((v * 100));
-				return (v > 2.25);
-			}
+		if (strncmp (map->name, core->bin->file, strlen (map->name)) == 0) {
+			continue;
+		}
+		if (r_regex_match (".*libc[.-]", "e", map->name)) {
+			found_glibc_map = true;
+			break;
 		}
 	}
+	// TODO: handle static binaries
+	if (!found_glibc_map) {
+		R_LOG_WARN ("resolve_glibc_version: no libc found in memory maps (cannot handle static binaries)");
+		return false;
+	}
+	// At this point we found a map in memory that _should_ be libc
+	version = GH (get_glibc_version) (core, map->file);
+	if (version != 0)	{
+		core->dbg->glibc_version = (int) round((version * 100));
+		core->dbg->glibc_version_d = version;
+		core->dbg->glibc_version_resolved = true;
+		char version_buffer[315] = {0};
+		// TODO: better way snprintf to 4 chars warns
+		// note: ‘snprintf’ output between 4 and 314 bytes into a destination of size 4
+		snprintf (version_buffer, sizeof (version_buffer)-1, "%.2f", version);
```

Comment:
Use r_str_newf instead of randomly choose 315 bytes

Reply:
315 was chosen because of the compiler warning that is in the comment above it, apparently that is the maximum size a double can have as string